option to show line with diagnostic and a marker (clang-style)

**Summary**

from @snsokolov :
Currently `verible-verilog-syntax` and `verible-verilog-lint` display diagnostics in the form:

FILE:LINE:COL: ...

but it might be nice to show the user the whole line with the offending text, plus a marker (clang-style diagnostics):

```
file.sv:5:11: error here...
  foo bar baz;
          ^
```

In other modes when the linter is being used in a service, and is expected to print only (machine parse-able) FILE:LINE:COL:, this would need to be disabled.

--- a/common/analysis/BUILD
+++ b/common/analysis/BUILD
@@ -128,6 +128,7 @@ cc_library(
         "//common/text:text_structure",
         "//common/text:token_info",
         "//common/text:token_stream_view",
+        "//common/util:spacer",
         "@com_google_absl//absl/status",
         "@com_google_absl//absl/strings",
     ],

--- a/common/analysis/file_analyzer.cc
+++ b/common/analysis/file_analyzer.cc
@@ -30,6 +30,7 @@
 #include "common/text/text_structure.h"
 #include "common/text/token_info.h"
 #include "common/text/token_stream_view.h"
+#include "common/util/spacer.h"
 
 namespace verible {
 

--- a/common/analysis/file_analyzer.cc
+++ b/common/analysis/file_analyzer.cc
@@ -136,15 +136,24 @@ std::vector<std::string> FileAnalyzer::TokenErrorMessages() const {
 }
 
 std::string FileAnalyzer::LinterTokenErrorMessage(
-    const RejectedToken& error_token) const {
+    const RejectedToken& error_token, bool diagnostic_context) const {
   const LineColumnMap& line_column_map = Data().GetLineColumnMap();
   const absl::string_view base_text = Data().Contents();
+  
   std::ostringstream output_stream;
   output_stream << filename_ << ':';
   if (!error_token.token_info.isEOF()) {
     const auto left = line_column_map(error_token.token_info.left(base_text));
     output_stream << left << ": " << error_token.phase << " error, rejected \""
                   << error_token.token_info.text() << "\" (syntax-error).";
+    if(diagnostic_context)
+    {
+      const auto& lines = Data().Lines();
+      if (left.line < static_cast<int>(lines.size())) {
+        output_stream << "\n" << lines[left.line] << std::endl;
+        output_stream << verible::Spacer(left.column) << "^";
+      }
+    }
   } else {
     const int file_size = base_text.length();
     const auto end = line_column_map(file_size);
@@ -158,11 +167,11 @@ std::string FileAnalyzer::LinterTokenErrorMessage(
   return output_stream.str();
 }
 
-std::vector<std::string> FileAnalyzer::LinterTokenErrorMessages() const {
+std::vector<std::string> FileAnalyzer::LinterTokenErrorMessages(bool diagnostic_context) const {
   std::vector<std::string> messages;
   messages.reserve(rejected_tokens_.size());
   for (const auto& rejected_token : rejected_tokens_) {
-    messages.push_back(LinterTokenErrorMessage(rejected_token));
+    messages.push_back(LinterTokenErrorMessage(rejected_token, diagnostic_context));
   }
   return messages;
 }

--- a/common/analysis/file_analyzer.cc
+++ b/common/analysis/file_analyzer.cc
@@ -139,15 +139,14 @@ std::string FileAnalyzer::LinterTokenErrorMessage(
     const RejectedToken& error_token, bool diagnostic_context) const {
   const LineColumnMap& line_column_map = Data().GetLineColumnMap();
   const absl::string_view base_text = Data().Contents();
-  
+
   std::ostringstream output_stream;
   output_stream << filename_ << ':';
   if (!error_token.token_info.isEOF()) {
     const auto left = line_column_map(error_token.token_info.left(base_text));
     output_stream << left << ": " << error_token.phase << " error, rejected \""
                   << error_token.token_info.text() << "\" (syntax-error).";
-    if(diagnostic_context)
-    {
+    if (diagnostic_context) {
       const auto& lines = Data().Lines();
       if (left.line < static_cast<int>(lines.size())) {
         output_stream << "\n" << lines[left.line] << std::endl;
@@ -167,11 +166,13 @@ std::string FileAnalyzer::LinterTokenErrorMessage(
   return output_stream.str();
 }
 
-std::vector<std::string> FileAnalyzer::LinterTokenErrorMessages(bool diagnostic_context) const {
+std::vector<std::string> FileAnalyzer::LinterTokenErrorMessages(
+    bool diagnostic_context) const {
   std::vector<std::string> messages;
   messages.reserve(rejected_tokens_.size());
   for (const auto& rejected_token : rejected_tokens_) {
-    messages.push_back(LinterTokenErrorMessage(rejected_token, diagnostic_context));
+    messages.push_back(
+        LinterTokenErrorMessage(rejected_token, diagnostic_context));
   }
   return messages;
 }

--- a/common/analysis/file_analyzer.h
+++ b/common/analysis/file_analyzer.h
@@ -94,8 +94,14 @@ class FileAnalyzer : public TextStructure {
   std::vector<std::string> TokenErrorMessages() const;
 
   // Diagnostic message for rejected tokens for linter.
+  // Second argument is the show_context option. When enabled
+  // additional diagnostic line is concatenated to an error message 
+  // with marker that points to vulnerable token
   std::string LinterTokenErrorMessage(const RejectedToken&, bool) const;
 
+  // First argument is the show_context option. When enabled
+  // additional diagnostic line is concatenated to an error message 
+  // with marker that points to vulnerable token
   std::vector<std::string> LinterTokenErrorMessages(bool) const;
 
   const std::vector<RejectedToken>& GetRejectedTokens() const {

--- a/common/analysis/file_analyzer.h
+++ b/common/analysis/file_analyzer.h
@@ -94,9 +94,9 @@ class FileAnalyzer : public TextStructure {
   std::vector<std::string> TokenErrorMessages() const;
 
   // Diagnostic message for rejected tokens for linter.
-  std::string LinterTokenErrorMessage(const RejectedToken&) const;
+  std::string LinterTokenErrorMessage(const RejectedToken&, bool) const;
 
-  std::vector<std::string> LinterTokenErrorMessages() const;
+  std::vector<std::string> LinterTokenErrorMessages(bool) const;
 
   const std::vector<RejectedToken>& GetRejectedTokens() const {
     return rejected_tokens_;

--- a/common/analysis/file_analyzer_test.cc
+++ b/common/analysis/file_analyzer_test.cc
@@ -71,6 +71,26 @@ TEST(FileAnalyzerTest, TokenErrorMessageSameLine) {
   }
 }
 
+// Verify that an error token on one line is reported correctly.
+TEST(FileAnalyzerTest, TokenErrorMessageSameLineWithContext) {
+  const std::string text("hello, world\nbye w0rld\n");
+  FakeFileAnalyzer analyzer(text, "hello.txt");
+  const TokenInfo error_token(1, analyzer.Data().Contents().substr(17, 5));
+  {
+    const auto message = analyzer.TokenErrorMessage(error_token);
+    EXPECT_EQ(message, "token: \"w0rld\" at 2:5-9");
+  }
+  {
+    constexpr bool with_diagnostic_contex = true;
+    const auto message = analyzer.LinterTokenErrorMessage(
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+    EXPECT_TRUE(
+        absl::StrContains(message,
+                          "hello.txt:2:5: syntax error, rejected \"w0rld\" "
+                          "(syntax-error).\nbye w0rld\n    ^"));
+  }
+}
+
 // Verify that an error token on one character is reported correctly.
 TEST(FileAnalyzerTest, TokenErrorMessageOneChar) {
   const std::string text("hello, world\nbye w0rld\n");
@@ -86,6 +106,25 @@ TEST(FileAnalyzerTest, TokenErrorMessageOneChar) {
         {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
     EXPECT_TRUE(absl::StrContains(
         message, "hello.txt:1:6: syntax error, rejected \",\""));
+  }
+}
+
+// Verify that an error token on one character is reported correctly.
+TEST(FileAnalyzerTest, TokenErrorMessageOneCharWithContext) {
+  const std::string text("hello, world\nbye w0rld\n");
+  FakeFileAnalyzer analyzer(text, "hello.txt");
+  const TokenInfo error_token(1, analyzer.Data().Contents().substr(5, 1));
+  {
+    const auto message = analyzer.TokenErrorMessage(error_token);
+    EXPECT_EQ(message, "token: \",\" at 1:6");
+  }
+  {
+    constexpr bool with_diagnostic_contex = true;
+    const auto message = analyzer.LinterTokenErrorMessage(
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+    EXPECT_TRUE(absl::StrContains(message,
+                                  "hello.txt:1:6: syntax error, rejected \",\" "
+                                  "(syntax-error).\nhello, world\n     ^"));
   }
 }
 
@@ -106,6 +145,27 @@ TEST(FileAnalyzerTest, TokenErrorMessageDifferentLine) {
         message, "hello.txt:1:8: syntax error, rejected \"world\nbye\""));
   }
 }
+
+// Verify that an error token spanning multiple lines is reported correctly.
+TEST(FileAnalyzerTest, TokenErrorMessageDifferentLineWithContext) {
+  const std::string text("hello, world\nbye w0rld\n");
+  FakeFileAnalyzer analyzer(text, "hello.txt");
+  const TokenInfo error_token(1, analyzer.Data().Contents().substr(7, 9));
+  {
+    const auto message = analyzer.TokenErrorMessage(error_token);
+    EXPECT_EQ(message, "token: \"world\nbye\" at 1:8-2:3");
+  }
+  {
+    constexpr bool with_diagnostic_contex = true;
+    const auto message = analyzer.LinterTokenErrorMessage(
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+    EXPECT_TRUE(absl::StrContains(
+        message,
+        "hello.txt:1:8: syntax error, rejected \"world\nbye\" "
+        "(syntax-error).\nhello, world\n       ^"));
+  }
+}
+
 //
 // Verify that an error at EOF reported correctly.
 TEST(FileAnalyzerTest, TokenErrorMessageEOF) {
@@ -118,6 +178,25 @@ TEST(FileAnalyzerTest, TokenErrorMessageEOF) {
   }
   {
     constexpr bool with_diagnostic_contex = false;
+    const auto message = analyzer.LinterTokenErrorMessage(
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+    EXPECT_TRUE(absl::StrContains(
+        message, "unbalanced.txt:3:1: syntax error (unexpected EOF)"));
+  }
+}
+
+//
+// Verify that an error at EOF reported correctly.
+TEST(FileAnalyzerTest, TokenErrorMessageEOFWithContext) {
+  const std::string text("hello, world\nbye w0rld (\n");
+  const TokenInfo error_token(TokenInfo::EOFToken());
+  FakeFileAnalyzer analyzer(text, "unbalanced.txt");
+  {
+    const auto message = analyzer.TokenErrorMessage(error_token);
+    EXPECT_EQ(message, "token: <<EOF>> at 3:1");
+  }
+  {
+    constexpr bool with_diagnostic_contex = true;
     const auto message = analyzer.LinterTokenErrorMessage(
         {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
     EXPECT_TRUE(absl::StrContains(

--- a/common/analysis/file_analyzer_test.cc
+++ b/common/analysis/file_analyzer_test.cc
@@ -63,8 +63,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageSameLine) {
     EXPECT_EQ(message, "token: \"w0rld\" at 2:5-9");
   }
   {
+    constexpr bool with_diagnostic_contex = false;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase});
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
     EXPECT_TRUE(absl::StrContains(
         message, "hello.txt:2:5: syntax error, rejected \"w0rld\""));
   }
@@ -80,8 +81,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageOneChar) {
     EXPECT_EQ(message, "token: \",\" at 1:6");
   }
   {
+    constexpr bool with_diagnostic_contex = false;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase});
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
     EXPECT_TRUE(absl::StrContains(
         message, "hello.txt:1:6: syntax error, rejected \",\""));
   }
@@ -97,8 +99,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageDifferentLine) {
     EXPECT_EQ(message, "token: \"world\nbye\" at 1:8-2:3");
   }
   {
+    constexpr bool with_diagnostic_contex = false;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase});
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
     EXPECT_TRUE(absl::StrContains(
         message, "hello.txt:1:8: syntax error, rejected \"world\nbye\""));
   }
@@ -114,8 +117,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageEOF) {
     EXPECT_EQ(message, "token: <<EOF>> at 3:1");
   }
   {
+    constexpr bool with_diagnostic_contex = false;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase});
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
     EXPECT_TRUE(absl::StrContains(
         message, "unbalanced.txt:3:1: syntax error (unexpected EOF)"));
   }

--- a/common/analysis/file_analyzer_test.cc
+++ b/common/analysis/file_analyzer_test.cc
@@ -63,9 +63,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageSameLine) {
     EXPECT_EQ(message, "token: \"w0rld\" at 2:5-9");
   }
   {
-    constexpr bool with_diagnostic_contex = false;
+    constexpr bool with_diagnostic_context = false;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(absl::StrContains(
         message, "hello.txt:2:5: syntax error, rejected \"w0rld\""));
   }
@@ -81,9 +81,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageSameLineWithContext) {
     EXPECT_EQ(message, "token: \"w0rld\" at 2:5-9");
   }
   {
-    constexpr bool with_diagnostic_contex = true;
+    constexpr bool with_diagnostic_context = true;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(
         absl::StrContains(message,
                           "hello.txt:2:5: syntax error, rejected \"w0rld\" "
@@ -101,9 +101,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageOneChar) {
     EXPECT_EQ(message, "token: \",\" at 1:6");
   }
   {
-    constexpr bool with_diagnostic_contex = false;
+    constexpr bool with_diagnostic_context = false;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(absl::StrContains(
         message, "hello.txt:1:6: syntax error, rejected \",\""));
   }
@@ -119,9 +119,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageOneCharWithContext) {
     EXPECT_EQ(message, "token: \",\" at 1:6");
   }
   {
-    constexpr bool with_diagnostic_contex = true;
+    constexpr bool with_diagnostic_context = true;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(absl::StrContains(message,
                                   "hello.txt:1:6: syntax error, rejected \",\" "
                                   "(syntax-error).\nhello, world\n     ^"));
@@ -138,9 +138,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageDifferentLine) {
     EXPECT_EQ(message, "token: \"world\nbye\" at 1:8-2:3");
   }
   {
-    constexpr bool with_diagnostic_contex = false;
+    constexpr bool with_diagnostic_context = false;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(absl::StrContains(
         message, "hello.txt:1:8: syntax error, rejected \"world\nbye\""));
   }
@@ -156,9 +156,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageDifferentLineWithContext) {
     EXPECT_EQ(message, "token: \"world\nbye\" at 1:8-2:3");
   }
   {
-    constexpr bool with_diagnostic_contex = true;
+    constexpr bool with_diagnostic_context = true;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(absl::StrContains(
         message,
         "hello.txt:1:8: syntax error, rejected \"world\nbye\" "
@@ -177,9 +177,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageEOF) {
     EXPECT_EQ(message, "token: <<EOF>> at 3:1");
   }
   {
-    constexpr bool with_diagnostic_contex = false;
+    constexpr bool with_diagnostic_context = false;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(absl::StrContains(
         message, "unbalanced.txt:3:1: syntax error (unexpected EOF)"));
   }
@@ -196,9 +196,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageEOFWithContext) {
     EXPECT_EQ(message, "token: <<EOF>> at 3:1");
   }
   {
-    constexpr bool with_diagnostic_contex = true;
+    constexpr bool with_diagnostic_context = true;
     const auto message = analyzer.LinterTokenErrorMessage(
-        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_contex);
+        {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(absl::StrContains(
         message, "unbalanced.txt:3:1: syntax error (unexpected EOF)"));
   }

--- a/common/analysis/file_analyzer_test.cc
+++ b/common/analysis/file_analyzer_test.cc
@@ -87,7 +87,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageSameLineWithContext) {
     EXPECT_TRUE(
         absl::StrContains(message,
                           "hello.txt:2:5: syntax error, rejected \"w0rld\" "
-                          "(syntax-error).\nbye w0rld\n    ^"));
+                          "(syntax-error).\n"
+                          "bye w0rld\n"
+                          "    ^"));
   }
 }
 
@@ -124,7 +126,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageOneCharWithContext) {
         {error_token, AnalysisPhase::kParsePhase}, with_diagnostic_context);
     EXPECT_TRUE(absl::StrContains(message,
                                   "hello.txt:1:6: syntax error, rejected \",\" "
-                                  "(syntax-error).\nhello, world\n     ^"));
+                                  "(syntax-error).\n"
+                                  "hello, world\n"
+                                  "     ^"));
   }
 }
 
@@ -162,7 +166,9 @@ TEST(FileAnalyzerTest, TokenErrorMessageDifferentLineWithContext) {
     EXPECT_TRUE(absl::StrContains(
         message,
         "hello.txt:1:8: syntax error, rejected \"world\nbye\" "
-        "(syntax-error).\nhello, world\n       ^"));
+        "(syntax-error).\n"
+        "hello, world\n"
+        "       ^"));
   }
 }
 

--- a/common/analysis/lint_rule_status.cc
+++ b/common/analysis/lint_rule_status.cc
@@ -16,6 +16,7 @@
 
 #include <algorithm>
 #include <functional>
+#include <iostream>
 #include <iterator>
 #include <ostream>
 #include <string>
@@ -72,8 +73,10 @@ struct LintViolationWithStatus {
   }
 };
 
-static std::set<LintViolationWithStatus> PrepareViolations(
-    const std::vector<LintRuleStatus>& statuses) {
+void LintStatusFormatter::FormatLintRuleStatuses(
+    std::ostream* stream, const std::vector<LintRuleStatus>& statuses,
+    absl::string_view base, absl::string_view path,
+    const std::vector<absl::string_view>& lines) const {
   std::set<LintViolationWithStatus> violations;
 
   // TODO(fangism): rewrite as a linear time merge of pre-ordered sub-sequences
@@ -82,33 +85,16 @@ static std::set<LintViolationWithStatus> PrepareViolations(
       violations.insert(LintViolationWithStatus(&violation, &status));
     }
   }
-  return violations;
-}
-
-void LintStatusFormatter::FormatLintRuleStatuses(
-    std::ostream* stream, const std::vector<LintRuleStatus>& statuses,
-    absl::string_view base, absl::string_view path) const {
-  auto violations = PrepareViolations(statuses);
-  for (auto violation : violations) {
-    FormatViolation(stream, *violation.violation, base, path,
-                    violation.status->url, violation.status->lint_rule_name);
-    *stream << std::endl;
-  }
-}
-
-void LintStatusFormatter::FormatLintRuleStatuses(
-    std::ostream* stream, const std::vector<LintRuleStatus>& statuses,
-    absl::string_view base, absl::string_view path,
-    const std::vector<absl::string_view>& lines) const {
-  auto violations = PrepareViolations(statuses);
 
   for (auto violation : violations) {
     FormatViolation(stream, *violation.violation, base, path,
                     violation.status->url, violation.status->lint_rule_name);
     *stream << std::endl;
     auto cursor = line_column_map_(violation.violation->token.left(base));
-    *stream << lines[cursor.line] << std::endl;
-    *stream << verible::Spacer(cursor.column) << "^" << std::endl;
+    if (cursor.line < lines.size()) {
+      *stream << lines[cursor.line] << std::endl;
+      *stream << verible::Spacer(cursor.column) << "^" << std::endl;
+    }
   }
 }
 

--- a/common/analysis/lint_rule_status.cc
+++ b/common/analysis/lint_rule_status.cc
@@ -28,6 +28,7 @@
 #include "common/text/syntax_tree_context.h"
 #include "common/text/token_info.h"
 #include "common/text/tree_utils.h"
+#include "common/util/spacer.h"
 
 namespace verible {
 
@@ -71,9 +72,8 @@ struct LintViolationWithStatus {
   }
 };
 
-void LintStatusFormatter::FormatLintRuleStatuses(
-    std::ostream* stream, const std::vector<LintRuleStatus>& statuses,
-    absl::string_view base, absl::string_view path) const {
+static std::set<LintViolationWithStatus> PrepareViolations(
+    const std::vector<LintRuleStatus>& statuses) {
   std::set<LintViolationWithStatus> violations;
 
   // TODO(fangism): rewrite as a linear time merge of pre-ordered sub-sequences
@@ -82,11 +82,33 @@ void LintStatusFormatter::FormatLintRuleStatuses(
       violations.insert(LintViolationWithStatus(&violation, &status));
     }
   }
+  return violations;
+}
+
+void LintStatusFormatter::FormatLintRuleStatuses(
+    std::ostream* stream, const std::vector<LintRuleStatus>& statuses,
+    absl::string_view base, absl::string_view path) const {
+  auto violations = PrepareViolations(statuses);
+  for (auto violation : violations) {
+    FormatViolation(stream, *violation.violation, base, path,
+                    violation.status->url, violation.status->lint_rule_name);
+    *stream << std::endl;
+  }
+}
+
+void LintStatusFormatter::FormatLintRuleStatuses(
+    std::ostream* stream, const std::vector<LintRuleStatus>& statuses,
+    absl::string_view base, absl::string_view path,
+    const std::vector<absl::string_view>& lines) const {
+  auto violations = PrepareViolations(statuses);
 
   for (auto violation : violations) {
     FormatViolation(stream, *violation.violation, base, path,
                     violation.status->url, violation.status->lint_rule_name);
     *stream << std::endl;
+    auto cursor = line_column_map_(violation.violation->token.left(base));
+    *stream << lines[cursor.line] << std::endl;
+    *stream << verible::Spacer(cursor.column) << "^" << std::endl;
   }
 }
 

--- a/common/analysis/lint_rule_status.cc
+++ b/common/analysis/lint_rule_status.cc
@@ -91,7 +91,7 @@ void LintStatusFormatter::FormatLintRuleStatuses(
                     violation.status->url, violation.status->lint_rule_name);
     *stream << std::endl;
     auto cursor = line_column_map_(violation.violation->token.left(base));
-    if (cursor.line < lines.size()) {
+    if (cursor.line < static_cast<int>(lines.size())) {
       *stream << lines[cursor.line] << std::endl;
       *stream << verible::Spacer(cursor.column) << "^" << std::endl;
     }

--- a/common/analysis/lint_rule_status.h
+++ b/common/analysis/lint_rule_status.h
@@ -122,7 +122,8 @@ class LintStatusFormatter {
                             absl::string_view base,
                             absl::string_view path) const;
 
-  // Formats, sorts and outputs status to stream.
+  // Formats, sorts and outputs status to stream with additional vulnerable code
+  // line printed when enabled.
   // The violations contained in the statuses are sorted by their occurrence
   // in the code and are not grouped by the status object.
   // Path is the file path of original file. This is needed because it is not
@@ -132,18 +133,8 @@ class LintStatusFormatter {
   void FormatLintRuleStatuses(std::ostream* stream,
                               const std::vector<LintRuleStatus>& statuses,
                               absl::string_view base,
-                              absl::string_view path) const;
-
-  // Formats, sorts and outputs status to stream with additional vulnerable code
-  // line printed when enabled. The violations contained in the statuses are
-  // sorted by their occurrence in the code and are not grouped by the status
-  // object. Path is the file path of original file. This is needed because it
-  // is not contained in status. Base is the string_view of the entire contents,
-  // used only for byte offset calculation.
-  void FormatLintRuleStatuses(
-      std::ostream* stream, const std::vector<LintRuleStatus>& statuses,
-      absl::string_view base, absl::string_view path,
-      const std::vector<absl::string_view>& lines) const;
+                              absl::string_view path,
+                              const std::vector<absl::string_view>& lines) const;
 
   // Formats and outputs violation on stream.
   // Path is file path of original file and url is a link to the ratified rule

--- a/common/analysis/lint_rule_status.h
+++ b/common/analysis/lint_rule_status.h
@@ -134,6 +134,17 @@ class LintStatusFormatter {
                               absl::string_view base,
                               absl::string_view path) const;
 
+  // Formats, sorts and outputs status to stream with additional vulnerable code
+  // line printed when enabled. The violations contained in the statuses are
+  // sorted by their occurrence in the code and are not grouped by the status
+  // object. Path is the file path of original file. This is needed because it
+  // is not contained in status. Base is the string_view of the entire contents,
+  // used only for byte offset calculation.
+  void FormatLintRuleStatuses(
+      std::ostream* stream, const std::vector<LintRuleStatus>& statuses,
+      absl::string_view base, absl::string_view path,
+      const std::vector<absl::string_view>& lines) const;
+
   // Formats and outputs violation on stream.
   // Path is file path of original file and url is a link to the ratified rule
   // that is being violated.

--- a/common/analysis/lint_rule_status_test.cc
+++ b/common/analysis/lint_rule_status_test.cc
@@ -144,7 +144,7 @@ TEST(LintRuleStatusFormatterTest, NoOutput) {
   RunLintStatusTest(test);
 }
 
-void RunLintStatusesTest(const LintStatusTest& test) {
+void RunLintStatusesTest(const LintStatusTest& test, bool show_context) {
   // Dummy tree so we have something for test cases to point at
   SymbolPtr root = Node();
 
@@ -172,8 +172,21 @@ void RunLintStatusesTest(const LintStatusTest& test) {
   std::ostringstream ss;
 
   LintStatusFormatter formatter(test.text);
-  formatter.FormatLintRuleStatuses(&ss, statuses, test.text, test.path);
-  auto result_parts = absl::StrSplit(ss.str(), '\n');
+  const std::vector<absl::string_view> lines;
+  if (!show_context) {
+    formatter.FormatLintRuleStatuses(&ss, statuses, test.text, test.path);
+  } else {
+    formatter.FormatLintRuleStatuses(&ss, statuses, test.text, test.path,
+                                     absl::StrSplit(test.text, '\n'));
+    std::cout << ss.str() << std::endl;
+  }
+  std::vector<std::string> result_parts;
+  if (!show_context)
+    result_parts = absl::StrSplit(ss.str(), '\n');
+  else {
+    result_parts = absl::StrSplit(ss.str(), "^\n");
+  }
+
   auto part_iterator = result_parts.begin();
 
   for (const auto& violation_test : test.violations) {
@@ -200,7 +213,29 @@ TEST(LintRuleStatusFormatterTest, MultipleStatusesSimpleOutput) {
        {"reason2", TokenInfo(dont_care_tag, text.substr(21, 4)),
         "some/path/to/somewhere.fvg:2:4: reason2 http://foobar [test-rule]"}}};
 
-  RunLintStatusesTest(test);
+  RunLintStatusesTest(test, false);
+}
+
+TEST(LintRuleStatusFormatterTestWithContext, MultipleStatusesSimpleOutput) {
+  SymbolPtr root = Node();
+  static const int dont_care_tag = 0;
+  constexpr absl::string_view text(
+      "This is some code\n"
+      "That you are looking at right now\n"
+      "It is nice code, make no mistake\n"
+      "Very nice");
+  LintStatusTest test = {
+      "test-rule",
+      "http://foobar",
+      "some/path/to/somewhere.fvg",
+      text,
+      {{"reason1", TokenInfo(dont_care_tag, text.substr(0, 5)),
+        "some/path/to/somewhere.fvg:1:1: reason1 http://foobar "
+        "[test-rule]\nThis is some code\n"},
+       {"reason2", TokenInfo(dont_care_tag, text.substr(21, 4)),
+        "some/path/to/somewhere.fvg:2:4: reason2 http://foobar "
+        "[test-rule]\nThat you are looking at right now\n   "}}};
+  RunLintStatusesTest(test, true);
 }
 
 }  // namespace

--- a/common/analysis/lint_rule_status_test.cc
+++ b/common/analysis/lint_rule_status_test.cc
@@ -174,7 +174,7 @@ void RunLintStatusesTest(const LintStatusTest& test, bool show_context) {
   LintStatusFormatter formatter(test.text);
   const std::vector<absl::string_view> lines;
   if (!show_context) {
-    formatter.FormatLintRuleStatuses(&ss, statuses, test.text, test.path);
+    formatter.FormatLintRuleStatuses(&ss, statuses, test.text, test.path, {});
   } else {
     formatter.FormatLintRuleStatuses(&ss, statuses, test.text, test.path,
                                      absl::StrSplit(test.text, '\n'));

--- a/verilog/analysis/verilog_analyzer_test.cc
+++ b/verilog/analysis/verilog_analyzer_test.cc
@@ -75,8 +75,9 @@ bool TreeContainsToken(const ConcreteSyntaxTree& tree, const TokenInfo& token) {
 
 void DiagnosticMessagesContainFilename(const VerilogAnalyzer& analyzer,
                                        absl::string_view filename) {
+  constexpr bool with_diagnostic_contex = false;
   const std::vector<std::string> syntax_error_messages(
-      analyzer.LinterTokenErrorMessages());
+      analyzer.LinterTokenErrorMessages(with_diagnostic_contex));
   for (const auto& message : syntax_error_messages) {
     EXPECT_TRUE(absl::StrContains(message, filename));
   }

--- a/verilog/analysis/verilog_analyzer_test.cc
+++ b/verilog/analysis/verilog_analyzer_test.cc
@@ -74,10 +74,10 @@ bool TreeContainsToken(const ConcreteSyntaxTree& tree, const TokenInfo& token) {
 }
 
 void DiagnosticMessagesContainFilename(const VerilogAnalyzer& analyzer,
-                                       absl::string_view filename) {
-  constexpr bool with_diagnostic_contex = false;
+                                       absl::string_view filename,
+                                       bool with_diagnostic_context) {
   const std::vector<std::string> syntax_error_messages(
-      analyzer.LinterTokenErrorMessages(with_diagnostic_contex));
+      analyzer.LinterTokenErrorMessages(with_diagnostic_context));
   for (const auto& message : syntax_error_messages) {
     EXPECT_TRUE(absl::StrContains(message, filename));
   }
@@ -103,7 +103,8 @@ TEST(AnalyzeVerilogLexerTest, RejectsBadId) {
   const auto& rejects = analyzer_ptr->GetRejectedTokens();
   EXPECT_THAT(rejects, SizeIs(1));
   EXPECT_EQ(rejects.front().phase, AnalysisPhase::kLexPhase);
-  DiagnosticMessagesContainFilename(*analyzer_ptr, "<noname>");
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<noname>", false);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<noname>", true);
 }
 
 // Tests that invalid macro identifier is rejected.
@@ -116,7 +117,8 @@ TEST(AnalyzeVerilogLexerTest, RejectsMacroBadId) {
   const auto& rejects = analyzer_ptr->GetRejectedTokens();
   EXPECT_THAT(rejects, SizeIs(1));
   EXPECT_EQ(rejects.front().phase, AnalysisPhase::kLexPhase);
-  DiagnosticMessagesContainFilename(*analyzer_ptr, "<noname>");
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<noname>", false);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<noname>", true);
 }
 
 // The following tests check that standalone Verilog expression parsing work.
@@ -160,7 +162,8 @@ TEST(AnalyzeVerilogExpressionTest, ParsesUnfinishedOp) {
   const auto& rejects = analyzer_ptr->GetRejectedTokens();
   EXPECT_THAT(rejects, SizeIs(1));
   EXPECT_EQ(rejects.front().phase, AnalysisPhase::kParsePhase);
-  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>");
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", false);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", true);
 }
 
 TEST(AnalyzeVerilogExpressionTest, Unbalanced) {
@@ -171,7 +174,8 @@ TEST(AnalyzeVerilogExpressionTest, Unbalanced) {
   const auto& rejects = analyzer_ptr->GetRejectedTokens();
   EXPECT_THAT(rejects, SizeIs(1));
   EXPECT_EQ(rejects.front().phase, AnalysisPhase::kParsePhase);
-  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>");
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", false);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", true);
 }
 
 TEST(AnalyzeVerilogExpressionTest, ParsesConcat) {
@@ -208,7 +212,8 @@ TEST(AnalyzeVerilogExpressionTest, RejectsModuleItemAttack) {
       << "got: [\n"
       << verible::SequenceFormatter(rejects, "\n") << "\n]";
   EXPECT_EQ(rejects.front().phase, AnalysisPhase::kParsePhase);
-  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>");
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", false);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", true);
 }
 
 // The following tests check that standalone Verilog module-body parsing works.
@@ -285,7 +290,8 @@ TEST(AnalyzeVerilogClassBodyTest, RejectsModuleItem) {
   const auto& first_reject = rejects.front();
   EXPECT_EQ(first_reject.phase, AnalysisPhase::kParsePhase);
   EXPECT_EQ(first_reject.token_info.text(), "initial");
-  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>");
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", false);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", true);
 }
 
 // The following tests check for package-body parsing.
@@ -315,7 +321,8 @@ TEST(AnalyzeVerilogClassBodyTest, RejectsWireDeclaration) {
   const auto& rejects = analyzer_ptr->GetRejectedTokens();
   EXPECT_THAT(rejects, SizeIs(1));
   EXPECT_EQ(rejects.front().phase, AnalysisPhase::kParsePhase);
-  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>");
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", false);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, "<file>", true);
 }
 
 TEST(AnalyzeVerilogLibraryMapTest, ParsesLibraryDeclaration) {
@@ -386,7 +393,8 @@ TEST(AnalyzeVerilogAutomaticMode, ModuleBodyModeSyntaxError) {
           filename);
   const auto status = ABSL_DIE_IF_NULL(analyzer_ptr)->ParseStatus();
   EXPECT_FALSE(status.ok());
-  DiagnosticMessagesContainFilename(*analyzer_ptr, filename);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, filename, false);
+  DiagnosticMessagesContainFilename(*analyzer_ptr, filename, true);
 }
 
 TEST(AnalyzeVerilogAutomaticMode, ClassBodyMode) {

--- a/verilog/analysis/verilog_linter.cc
+++ b/verilog/analysis/verilog_linter.cc
@@ -300,7 +300,7 @@ absl::Status VerilogLintTextStructure(std::ostream* stream,
     verible::LintStatusFormatter formatter(contents);
     if (!show_context) {
       formatter.FormatLintRuleStatuses(stream, linter_statuses, text_base,
-                                       filename);
+                                       filename, {});
     } else {
       formatter.FormatLintRuleStatuses(stream, linter_statuses, text_base,
                                        filename, text_structure.Lines());

--- a/verilog/analysis/verilog_linter.cc
+++ b/verilog/analysis/verilog_linter.cc
@@ -100,9 +100,8 @@ int LintOneFile(std::ostream* stream, absl::string_view filename,
     const auto lex_status = ABSL_DIE_IF_NULL(analyzer)->LexStatus();
     const auto parse_status = analyzer->ParseStatus();
     if (!lex_status.ok() || !parse_status.ok()) {
-      constexpr bool with_diagnostic_context = false;
       const std::vector<std::string> syntax_error_messages(
-          analyzer->LinterTokenErrorMessages(with_diagnostic_context));
+          analyzer->LinterTokenErrorMessages(show_context));
       for (const auto& message : syntax_error_messages) {
         *stream << message << std::endl;
       }

--- a/verilog/analysis/verilog_linter.cc
+++ b/verilog/analysis/verilog_linter.cc
@@ -100,8 +100,9 @@ int LintOneFile(std::ostream* stream, absl::string_view filename,
     const auto lex_status = ABSL_DIE_IF_NULL(analyzer)->LexStatus();
     const auto parse_status = analyzer->ParseStatus();
     if (!lex_status.ok() || !parse_status.ok()) {
+      constexpr bool with_diagnostic_context = false;
       const std::vector<std::string> syntax_error_messages(
-          analyzer->LinterTokenErrorMessages());
+          analyzer->LinterTokenErrorMessages(with_diagnostic_context));
       for (const auto& message : syntax_error_messages) {
         *stream << message << std::endl;
       }

--- a/verilog/analysis/verilog_linter.cc
+++ b/verilog/analysis/verilog_linter.cc
@@ -83,7 +83,7 @@ using verible::TokenInfo;
 //  2..: other fatal issues such as file not found.
 int LintOneFile(std::ostream* stream, absl::string_view filename,
                 const LinterConfiguration& config, bool check_syntax,
-                bool parse_fatal, bool lint_fatal) {
+                bool parse_fatal, bool lint_fatal, bool show_context) {
   std::string content;
   const absl::Status content_status =
       verible::file::GetContents(filename, &content);
@@ -115,8 +115,9 @@ int LintOneFile(std::ostream* stream, absl::string_view filename,
 
   // Analyze the parsed structure for lint violations.
   std::ostringstream lint_stream;
-  const absl::Status lint_status = VerilogLintTextStructure(
-      &lint_stream, std::string(filename), content, config, analyzer->Data());
+  const absl::Status lint_status =
+      VerilogLintTextStructure(&lint_stream, std::string(filename), content,
+                               config, analyzer->Data(), show_context);
   if (!lint_status.ok()) {
     // Something went wrong with running the lint analysis itself.
     LOG(ERROR) << "Fatal error: " << lint_status.message();
@@ -271,7 +272,8 @@ absl::Status VerilogLintTextStructure(std::ostream* stream,
                                       const std::string& filename,
                                       const std::string& contents,
                                       const LinterConfiguration& config,
-                                      const TextStructureView& text_structure) {
+                                      const TextStructureView& text_structure,
+                                      bool show_context) {
   // Create the linter, add rules, and run it.
   VerilogLinter linter;
   const absl::Status configuration_status = linter.Configure(config, filename);
@@ -296,8 +298,13 @@ absl::Status VerilogLintTextStructure(std::ostream* stream,
     VLOG(1) << "Lint Violations (" << total_violations << "): " << std::endl;
     // Output results to stream using formatter.
     verible::LintStatusFormatter formatter(contents);
-    formatter.FormatLintRuleStatuses(stream, linter_statuses, text_base,
-                                     filename);
+    if (!show_context) {
+      formatter.FormatLintRuleStatuses(stream, linter_statuses, text_base,
+                                       filename);
+    } else {
+      formatter.FormatLintRuleStatuses(stream, linter_statuses, text_base,
+                                       filename, text_structure.Lines());
+    }
   }
   return absl::OkStatus();
 }

--- a/verilog/analysis/verilog_linter.h
+++ b/verilog/analysis/verilog_linter.h
@@ -47,7 +47,7 @@ namespace verilog {
 // errors were found (syntax, lint), and anything else is a fatal error.
 int LintOneFile(std::ostream* stream, absl::string_view filename,
                 const LinterConfiguration& config, bool check_syntax,
-                bool parse_fatal, bool lint_fatal);
+                bool parse_fatal, bool lint_fatal, bool show_context);
 
 // VerilogLinter analyzes a TextStructureView of Verilog source code.
 // This uses syntax-tree based analyses and lexical token-stream analyses.
@@ -109,13 +109,14 @@ absl::Status AppendLinterConfigurationFromFile(
 //     and its sole purpose is to provide a translation from byte-offset to
 //     line-column in diagnostics.
 //   text_structure: contains the syntax tree that will be lint-analyzed.
+//   show_context: print additional line with vulnerable code
 //
 // Returns:
 //   absl::Status that reflects whether linter linter ran successfully.
 absl::Status VerilogLintTextStructure(
     std::ostream* stream, const std::string& filename,
     const std::string& contents, const LinterConfiguration& config,
-    const verible::TextStructureView& text_structure);
+    const verible::TextStructureView& text_structure, bool show_context);
 
 // Prints the rule, description and default_enabled.
 absl::Status PrintRuleInfo(std::ostream*,

--- a/verilog/analysis/verilog_linter_test.cc
+++ b/verilog/analysis/verilog_linter_test.cc
@@ -66,8 +66,8 @@ class LintOneFileTest : public DefaultLinterConfigTestFixture,
 // Tests that nonexistent file is handled as a fatal error.
 TEST_F(LintOneFileTest, FileNotFound) {
   std::ostringstream output;
-  const int exit_code =
-      LintOneFile(&output, "FileNotFound.sv", config_, true, false, false);
+  const int exit_code = LintOneFile(&output, "FileNotFound.sv", config_, true,
+                                    false, false, false);
   EXPECT_EQ(exit_code, 2);
 }
 
@@ -81,11 +81,20 @@ TEST_F(LintOneFileTest, LintCleanFiles) {
   };
   for (const auto test_code : kTestCases) {
     const ScopedTestFile temp_file(testing::TempDir(), test_code);
-    std::ostringstream output;
-    const int exit_code =
-        LintOneFile(&output, temp_file.filename(), config_, true, false, false);
-    EXPECT_EQ(exit_code, 0);
-    EXPECT_TRUE(output.str().empty());  // silence
+    {
+      std::ostringstream output;
+      const int exit_code = LintOneFile(&output, temp_file.filename(), config_,
+                                        true, false, false, false);
+      EXPECT_EQ(exit_code, 0);
+      EXPECT_TRUE(output.str().empty());  // silence
+    }
+    {  // enable additional error context printing
+      std::ostringstream output;
+      const int exit_code = LintOneFile(&output, temp_file.filename(), config_,
+                                        true, false, false, true);
+      EXPECT_EQ(exit_code, 0);
+      EXPECT_TRUE(output.str().empty());  // silence
+    }
   }
 }
 
@@ -101,21 +110,28 @@ TEST_F(LintOneFileTest, SyntaxError) {
     {  // continue even with syntax error
       std::ostringstream output;
       const int exit_code = LintOneFile(&output, temp_file.filename(), config_,
-                                        true, false, false);
+                                        true, false, false, false);
+      EXPECT_EQ(exit_code, 0);
+      EXPECT_FALSE(output.str().empty());
+    }
+    {  // continue even with syntax error with additional error context
+      std::ostringstream output;
+      const int exit_code = LintOneFile(&output, temp_file.filename(), config_,
+                                        true, false, false, true);
       EXPECT_EQ(exit_code, 0);
       EXPECT_FALSE(output.str().empty());
     }
     {  // abort on syntax error
       std::ostringstream output;
       const int exit_code = LintOneFile(&output, temp_file.filename(), config_,
-                                        true, true, false);
+                                        true, true, false, false);
       EXPECT_EQ(exit_code, 1);
       EXPECT_FALSE(output.str().empty());
     }
     {  // ignore syntax error
       std::ostringstream output;
       const int exit_code = LintOneFile(&output, temp_file.filename(), config_,
-                                        false, false, false);
+                                        false, false, false, false);
       EXPECT_EQ(exit_code, 0);
       EXPECT_TRUE(output.str().empty());  // silence
     }
@@ -133,14 +149,14 @@ TEST_F(LintOneFileTest, LintError) {
     {  // continue even with lint error
       std::ostringstream output;
       const int exit_code = LintOneFile(&output, temp_file.filename(), config_,
-                                        true, false, false);
+                                        true, false, false, false);
       EXPECT_EQ(exit_code, 0) << "output:\n" << output.str();
       EXPECT_FALSE(output.str().empty());
     }
     {  // abort on lint error
       std::ostringstream output;
       const int exit_code = LintOneFile(&output, temp_file.filename(), config_,
-                                        true, false, true);
+                                        true, false, true, false);
       EXPECT_EQ(exit_code, 1) << "output:\n" << output.str();
       EXPECT_FALSE(output.str().empty());
     }
@@ -172,7 +188,7 @@ class VerilogLinterTest : public DefaultLinterConfigTestFixture,
     // lint success, so as long as we have a syntax tree (even if there
     // are errors), run the lint checks.
     const absl::Status lint_status = VerilogLintTextStructure(
-        &diagnostics, filename, content, config_, analyzer->Data());
+        &diagnostics, filename, content, config_, analyzer->Data(), false);
     return {lint_status, diagnostics.str()};
   }
 };

--- a/verilog/analysis/verilog_linter_test.cc
+++ b/verilog/analysis/verilog_linter_test.cc
@@ -178,10 +178,17 @@ class VerilogLinterTest : public DefaultLinterConfigTestFixture,
     const absl::Status status = ABSL_DIE_IF_NULL(analyzer)->Analyze();
     std::ostringstream diagnostics;
     if (!status.ok()) {
-      constexpr bool with_diagnostic_contex = false;
+      bool with_diagnostic_contex = false;
       const std::vector<std::string> syntax_error_messages(
           analyzer->LinterTokenErrorMessages(with_diagnostic_contex));
       for (const auto& message : syntax_error_messages) {
+        diagnostics << message << std::endl;
+      }
+
+      with_diagnostic_contex = true;
+      const std::vector<std::string> syntax_error_messages_with_context(
+          analyzer->LinterTokenErrorMessages(with_diagnostic_contex));
+      for (const auto& message : syntax_error_messages_with_context) {
         diagnostics << message << std::endl;
       }
     }

--- a/verilog/analysis/verilog_linter_test.cc
+++ b/verilog/analysis/verilog_linter_test.cc
@@ -178,8 +178,9 @@ class VerilogLinterTest : public DefaultLinterConfigTestFixture,
     const absl::Status status = ABSL_DIE_IF_NULL(analyzer)->Analyze();
     std::ostringstream diagnostics;
     if (!status.ok()) {
+      constexpr bool with_diagnostic_contex = false;
       const std::vector<std::string> syntax_error_messages(
-          analyzer->LinterTokenErrorMessages());
+          analyzer->LinterTokenErrorMessages(with_diagnostic_contex));
       for (const auto& message : syntax_error_messages) {
         diagnostics << message << std::endl;
       }

--- a/verilog/formatting/formatter.cc
+++ b/verilog/formatting/formatter.cc
@@ -205,8 +205,9 @@ Status FormatVerilog(absl::string_view text, absl::string_view filename,
     const auto parse_status = analyzer->ParseStatus();
     if (!lex_status.ok() || !parse_status.ok()) {
       std::ostringstream errstream;
+      constexpr bool with_diagnostic_context = false;
       const std::vector<std::string> syntax_error_messages(
-          analyzer->LinterTokenErrorMessages());
+          analyzer->LinterTokenErrorMessages(with_diagnostic_context));
       for (const auto& message : syntax_error_messages) {
         errstream << message << std::endl;
       }

--- a/verilog/formatting/tree_unwrapper_test.cc
+++ b/verilog/formatting/tree_unwrapper_test.cc
@@ -239,8 +239,9 @@ class TreeUnwrapperTest : public ::testing::Test {
     // Since source code is required to be valid, this error-handling is just
     // to help debug the test case construction
     if (!status.ok()) {
+      constexpr bool with_diagnostic_contex = false;
       const std::vector<std::string> syntax_error_messages(
-          analyzer_->LinterTokenErrorMessages());
+          analyzer_->LinterTokenErrorMessages(with_diagnostic_contex));
       for (const auto& message : syntax_error_messages) {
         std::cout << message << std::endl;
       }

--- a/verilog/formatting/tree_unwrapper_test.cc
+++ b/verilog/formatting/tree_unwrapper_test.cc
@@ -239,9 +239,9 @@ class TreeUnwrapperTest : public ::testing::Test {
     // Since source code is required to be valid, this error-handling is just
     // to help debug the test case construction
     if (!status.ok()) {
-      constexpr bool with_diagnostic_contex = false;
+      constexpr bool with_diagnostic_context = false;
       const std::vector<std::string> syntax_error_messages(
-          analyzer_->LinterTokenErrorMessages(with_diagnostic_contex));
+          analyzer_->LinterTokenErrorMessages(with_diagnostic_context));
       for (const auto& message : syntax_error_messages) {
         std::cout << message << std::endl;
       }

--- a/verilog/tools/lint/lint_tool_test.sh
+++ b/verilog/tools/lint/lint_tool_test.sh
@@ -105,6 +105,35 @@ status="$?"
   exit 1
 }
 
+DIAGNOSTIC_OUTPUT="${TEST_TMPDIR}/expected-diagnostic.out"
+cat > "${DIAGNOSTIC_OUTPUT}" <<EOF
+${TEST_TMPDIR}/syntax-error.sv:2:1: syntax error, rejected "endclass" (syntax-error).
+endclass
+^
+EOF
+
+"$lint_tool" "$TEST_FILE" --show_diagnostic_context > "${MY_OUTPUT_FILE}.out"
+
+diff "${DIAGNOSTIC_OUTPUT}" "${MY_OUTPUT_FILE}.out"
+status="$?"
+[[ $status == 0 ]] || {
+  echo "Expected exit code 0, but got $status"
+  exit 1
+}
+
+cat > "${DIAGNOSTIC_OUTPUT}" <<EOF
+${TEST_TMPDIR}/syntax-error.sv:2:1: syntax error, rejected "endclass" (syntax-error).
+EOF
+
+"$lint_tool" "$TEST_FILE" > "${MY_OUTPUT_FILE}.out"
+
+diff "${DIAGNOSTIC_OUTPUT}" "${MY_OUTPUT_FILE}.out"
+status="$?"
+[[ $status == 0 ]] || {
+  echo "Expected exit code 0, but got $status"
+  exit 1
+}
+
 echo "=== Test --parse_fatal"
 "$lint_tool" "$TEST_FILE" --parse_fatal > /dev/null 2> "${MY_OUTPUT_FILE}.err"
 

--- a/verilog/tools/lint/lint_tool_test.sh
+++ b/verilog/tools/lint/lint_tool_test.sh
@@ -96,6 +96,15 @@ status="$?"
   exit 1
 }
 
+echo "=== Test --show_diagnostic_context"
+"$lint_tool" "$TEST_FILE" --show_diagnostic_context > /dev/null 2> "${MY_OUTPUT_FILE}.err"
+
+status="$?"
+[[ $status == 1 ]] || {
+  echo "Expected exit code 1, but got $status"
+  exit 1
+}
+
 echo "=== Test --parse_fatal"
 "$lint_tool" "$TEST_FILE" --parse_fatal > /dev/null 2> "${MY_OUTPUT_FILE}.err"
 

--- a/verilog/tools/lint/verilog_lint.cc
+++ b/verilog/tools/lint/verilog_lint.cc
@@ -50,6 +50,10 @@ ABSL_FLAG(
     "Markdown and exit immediately. Intended for the output to be written "
     "to a snippet of Markdown.");
 
+ABSL_FLAG(bool, show_diagnostic_context, false,
+          "prints an additional "
+          "line on which the diagnostic was found,"
+          "followed by a line with a position marker");
 // LINT.ThenChange(README.md)
 
 using verilog::LinterConfiguration;
@@ -83,7 +87,8 @@ int main(int argc, char** argv) {
     const int lint_status = verilog::LintOneFile(
         &std::cout, filename, config,  //
         absl::GetFlag(FLAGS_check_syntax), absl::GetFlag(FLAGS_parse_fatal),
-        absl::GetFlag(FLAGS_lint_fatal));
+        absl::GetFlag(FLAGS_lint_fatal),
+        absl::GetFlag(FLAGS_show_diagnostic_context));
     exit_status = std::max(lint_status, exit_status);
   }  // for each file
 

--- a/verilog/tools/syntax/verilog_syntax.cc
+++ b/verilog/tools/syntax/verilog_syntax.cc
@@ -103,6 +103,11 @@ ABSL_FLAG(
     bool, verifytree, false,
     "Verifies that all tokens are parsed into tree, prints unmatched tokens");
 
+ABSL_FLAG(bool, show_diagnostic_context, false,
+          "prints an additional "
+          "line on which the diagnostic was found,"
+          "followed by a line with a position marker");
+
 using verible::ConcreteSyntaxTree;
 using verible::ParserVerifier;
 using verible::TextStructureView;
@@ -151,6 +156,8 @@ static int AnalyzeOneFile(absl::string_view content, absl::string_view filename,
   const auto parse_status = analyzer->ParseStatus();
 
   if (!lex_status.ok() || !parse_status.ok()) {
+    const std::vector<std::string> syntax_error_messages(
+        analyzer->LinterTokenErrorMessages(absl::GetFlag(FLAGS_show_diagnostic_context)));
     const int error_limit = absl::GetFlag(FLAGS_error_limit);
     int error_count = 0;
     if (!absl::GetFlag(FLAGS_export_json)) {

--- a/verilog/tools/syntax/verilog_syntax.cc
+++ b/verilog/tools/syntax/verilog_syntax.cc
@@ -157,12 +157,14 @@ static int AnalyzeOneFile(absl::string_view content, absl::string_view filename,
 
   if (!lex_status.ok() || !parse_status.ok()) {
     const std::vector<std::string> syntax_error_messages(
-        analyzer->LinterTokenErrorMessages(absl::GetFlag(FLAGS_show_diagnostic_context)));
+        analyzer->LinterTokenErrorMessages(
+            absl::GetFlag(FLAGS_show_diagnostic_context)));
     const int error_limit = absl::GetFlag(FLAGS_error_limit);
     int error_count = 0;
     if (!absl::GetFlag(FLAGS_export_json)) {
       const std::vector<std::string> syntax_error_messages(
-          analyzer->LinterTokenErrorMessages());
+          analyzer->LinterTokenErrorMessages(
+              absl::GetFlag(FLAGS_show_diagnostic_context)));
       for (const auto& message : syntax_error_messages) {
         std::cout << message << std::endl;
         ++error_count;

